pkg/exemplars/exemplarspb/custom.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package exemplarspb
+
+import (
+	"encoding/json"
+	"math/big"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+)
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (m *Exemplar) UnmarshalJSON(b []byte) error {
+	v := struct {
+		Labels    labelpb.ZLabelSet
+		TimeStamp model.Time
+		Value     model.SampleValue
+	}{}
+	if err := json.Unmarshal(b, &v); err != nil {
+		return err
+	}
+
+	m.Labels = v.Labels
+	m.Ts = int64(v.TimeStamp)
+	m.Value = float64(v.Value)
+
+	return nil
+}
+
+// MarshalJSON implements json.Marshaler.
+func (m *Exemplar) MarshalJSON() ([]byte, error) {
+	v := struct {
+		Labels    labels.Labels     `json:"labels"`
+		TimeStamp model.Time        `json:"timestamp"`
+		Value     model.SampleValue `json:"value"`
+	}{
+		Labels:    labelpb.ZLabelsToPromLabels(m.Labels.Labels),
+		TimeStamp: model.Time(m.Ts),
+		Value:     model.SampleValue(m.Ts),
+	}
+	return json.Marshal(v)
+}
+
+func NewExemplarsResponse(e *ExemplarData) *ExemplarsResponse {
+	return &ExemplarsResponse{
+		Result: &ExemplarsResponse_Data{
+			Data: e,
+		},
+	}
+}
+
+func NewWarningExemplarsResponse(warning error) *ExemplarsResponse {
+	return &ExemplarsResponse{
+		Result: &ExemplarsResponse_Warning{
+			Warning: warning.Error(),
+		},
+	}
+}
+
+func (s1 *ExemplarData) Compare(s2 *ExemplarData) int {
+	if d := labels.Compare(s1.SeriesLabels.PromLabels(), s2.SeriesLabels.PromLabels()); d != 0 {
+		return d
+	}
+	return 0
```

Comment:
We can just use `labels.Compare` here, no need to use `if d := xxx; d != 0` pattern since we removed the `HasTS` field already.